File: android/measure-android/measure/src/test/java/sh/measure/android/layoutinspector/LayoutSnapshotThrottlerTest.kt
package sh.measure.android.layoutinspector

import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import sh.measure.android.utils.AndroidTimeProvider
import sh.measure.android.utils.TestClock
import java.time.Duration

class LayoutSnapshotThrottlerTest {
    private val testClock = TestClock.create()
    private val timeProvider = AndroidTimeProvider(testClock)
    private val throttler = LayoutSnapshotThrottler(timeProvider)

    @Test
    fun `allows the first snapshot`() {
        assertTrue(throttler.shouldTakeSnapshot())
    }

    @Test
    fun `blocks a snapshot taken too soon after the previous one`() {
        throttler.shouldTakeSnapshot()
        assertFalse(throttler.shouldTakeSnapshot())
    }

    @Test
    fun `allows a snapshot once the delay has elapsed`() {
        throttler.shouldTakeSnapshot()
        testClock.advance(Duration.ofMillis(751))
        assertTrue(throttler.shouldTakeSnapshot())
    }

    @Test
    fun `respects a custom delay`() {
        throttler.shouldTakeSnapshot(delayMs = 100)
        testClock.advance(Duration.ofMillis(50))
        assertFalse(throttler.shouldTakeSnapshot(delayMs = 100))
        testClock.advance(Duration.ofMillis(101))
        assertTrue(throttler.shouldTakeSnapshot(delayMs = 100))
    }
}

File: flutter/packages/measure_flutter/test/gestures/layout_snapshot_throttler_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:measure_flutter/src/gestures/layout_snapshot_throttler.dart';
import 'package:measure_flutter/src/time/time_provider.dart';

import '../utils/test_clock.dart';

void main() {
  group('LayoutSnapshotThrottler', () {
    late TestClock clock;
    late LayoutSnapshotThrottler throttler;

    setUp(() {
      clock = TestClock.create();
      throttler = LayoutSnapshotThrottler(FlutterTimeProvider(clock));
    });

    test('allows the first snapshot', () {
      expect(throttler.shouldTakeSnapshot(), isTrue);
    });

    test('blocks a snapshot taken too soon after the previous one', () {
      throttler.shouldTakeSnapshot();

      expect(throttler.shouldTakeSnapshot(), isFalse);
    });

    test('allows a snapshot once the delay has elapsed', () {
      throttler.shouldTakeSnapshot();
      clock.advance(const Duration(milliseconds: 751));

      expect(throttler.shouldTakeSnapshot(), isTrue);
    });

    test('respects a custom delay', () {
      throttler.shouldTakeSnapshot(delayMs: 100);
      clock.advance(const Duration(milliseconds: 50));

      expect(throttler.shouldTakeSnapshot(delayMs: 100), isFalse);

      clock.advance(const Duration(milliseconds: 101));

      expect(throttler.shouldTakeSnapshot(delayMs: 100), isTrue);
    });
  });
}
